llvm/lib/Target/AIE/AIESlotCounts.cpp

```diff
@@ -0,0 +1,71 @@
+//===- AIESlotCounts.cpp - SlotCount utilit  y                            ===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
+
+#include "AIESlotCounts.h"
+
+namespace llvm {
+namespace AIE {
+
+SlotCounts::SlotCounts(SlotBits Bits) {
+  while (Bits) {
+    assert(Size < MaxSlots);
+    Counts[Size] = Bits & 1;
+    Size++;
+    Bits >>= 1;
+  }
+}
+
+SlotCounts::SlotCounts(const SlotCounts &Org) : Size(Org.Size) {
+  for (int I = 0; I < Size; I++) {
+    Counts[I] = Org.Counts[I];
+  }
+}
+
+int SlotCounts::max() {
+  int Max = 0;
+  for (int I = 0; I < Size; I++) {
+    Max = std::max(Max, int(Counts[I]));
+  }
+  return Max;
+}
+
+SlotCounts &SlotCounts::operator+=(const SlotCounts &Other) {
+  // The common part
+  for (int I = 0; I < Size && I < Other.Size; I++) {
+    Counts[I] += Other.Counts[I];
+  }
+  // Any excess from the other
+  while (Size < Other.Size) {
+    Counts[Size++] = Other.Counts[Size];
```

Comment:
Hmm, `Counts[Size++]` will have the previous value of `Size` as index, but `Other.Counts[Size]` will use the old one? Maybe move the increment in a different statement to clear our any confusion?